### PR TITLE
Upgrade to `smoltcp` 0.9

### PR DESCRIPTION
**Pull request by mkeeter**
Pending the PR in `gateway-messages`, since otherwise interoperability breaks (we're using the `smoltcp` IPv6 address in our APIs)

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -117,7 +114,7 @@ zip = { version = "0.6", default-features = false, features = ["bzip2"] }
 
 # Oxide forks and repos
 dice-mfg-msgs = { git = "https://github.com/oxidecomputer/dice-util", default-features = false }
-gateway-messages = { git = "https://github.com/oxidecomputer/management-gateway-service", default-features = false, features = ["smoltcp"] }
+gateway-messages = { git = "https://github.com/oxidecomputer/management-gateway-service", default-features = false, features = ["smoltcp"], branch = "bump-smoltcp" }
```
Comment (mkeeter) [inline Cargo.toml:117]: NOTE: merge this first!

File: task/net/build.rs [hunk 2]
```diff
@@ -36,7 +36,7 @@ fn generate_net_config(config: &NetConfig) -> Result<()> {
         "{}",
         quote::quote! {
             use core::sync::atomic::{AtomicBool, Ordering};
-            use smoltcp::socket::{UdpPacketMetadata, UdpSocket, UdpSocketBuffer};
+            use smoltcp::socket::{udp};
```
Comment (cbiffle) [inline task/net/build.rs:39]: Nit, but: those curly braces are unnecessary.

File: task/net/src/server.rs [hunk 3]
```diff
@@ -20,9 +20,9 @@ use task_net_api::{
 
 use core::iter::zip;
 use heapless::Vec;
-use smoltcp::iface::{Interface, Neighbor, SocketHandle, SocketStorage};
-use smoltcp::socket::UdpSocket;
-use smoltcp::wire::{EthernetAddress, IpAddress, IpCidr, Ipv6Cidr};
+use smoltcp::iface::{Interface, SocketHandle, SocketStorage};
+use smoltcp::socket::udp::Socket as UdpSocket;
```
Comment (cbiffle) [inline task/net/src/server.rs:24]: (slight preference for just calling this `udp::Socket` in this file, as it makes googling the type easier)

File: task/net/src/server.rs [hunk 4]
```diff
@@ -561,7 +555,7 @@ where
                 self.client_waiting_to_send[socket_index] = false;
                 Ok(())
             }
-            Err(smoltcp::Error::Exhausted) => {
+            Err(smoltcp::socket::udp::SendError::BufferFull) => {
```
Comment (cbiffle) [inline task/net/src/server.rs:558]: (if you `use smoltcp::socket::udp` above to refer to `udp::Socket` then this becomes `udp::SendError` which'd be slightly more concise)

File: task/net/src/server.rs (hunk 3 above)
Comment (cbiffle) [inline task/net/src/server.rs:24]: (slight preference for importing `udp` and calling this type `udp::Socket` so it doesn't look like a new type)

File: task/net/build.rs (hunk 2 above)
Comment (cbiffle) [inline task/net/build.rs:39]: (nit: unnecessary curlies)

Comment (mkeeter) [reply task/net/build.rs]: Fixed!

File: task/net/src/server.rs (hunk 4 above)
Comment (mkeeter) [reply task/net/src/server.rs]: Good idea, done

File: task/net/build.rs (hunk 2 above)
Comment (mkeeter) [reply task/net/build.rs]: Duplicate comment (?), fixed!

File: task/net/src/server.rs (hunk 3 above)
Comment (mkeeter) [reply task/net/src/server.rs]: yup, seems wise

**Pull request merged**